packages/block-collab/src/use-block-collab/index.js

```diff
@@ -0,0 +1,149 @@
+/**
+ * External dependencies
+ */
+import * as yjs from 'yjs';
+import { WebrtcProvider } from 'y-webrtc';
+
+/**
+ * WordPress dependencies
+ */
+import { useState, useRef, useEffect } from '@wordpress/element';
+import { useDispatch } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import setYDocBlocks from './set-y-doc-blocks';
+import yDocBlocksToArray from './y-doc-blocks-to-array';
+
+const instances = {};
+
+/**
+ * Connects to the block collaboration room of the given ID,
+ * with the specified password, and returns a list of peers,
+ * as well as the shared blocks and a function to update them.
+ *
+ * @param {string} id       The ID.
+ * @param {string} password The password.
+ *
+ * @return {Object} The shared data.
+ */
+export default function useBlockCollab( id, password ) {
+	const [ peers, setPeers ] = useState( {} );
+	const [ setSelf, setSetSelf ] = useState( () => () => {} );
+
+	const [ blocks, _setBlocks ] = useState( [] );
+	const initialBlocksRef = useRef();
+	const [ setBlocks, setSetBlocks ] = useState( () => ( newBlocks ) => {
+		// Remember these local changes before initialization so that
+		// we can sync them later.
+		initialBlocksRef.current = newBlocks;
+		_setBlocks( newBlocks );
+	} );
+
+	const setAddBlockSelectionsState = useDispatch(
+		'core/block-collab/add-block-selections'
+	)?.setState;
+	useEffect( () => {
+		// We share instances between hook instances for the same document,
+		// because we need to reuse the same `yDoc` and provider.
+		// Note that the `localYjsChangeSymbol` does differ per hook instance,
+		// so that edits made through one sync to the others.
+		const instanceKey = `${ id }|${ password }`;
+		if ( ! instances[ instanceKey ] ) {
+			const yDoc = new yjs.Doc();
+			const yDocPeers = yDoc.getMap( 'peers' );
```

Comment:
I suggest using the Awareness protocol instead for showing who is currently working on the document. https://github.com/yjs/y-protocols/

It is a dedicated CRDT for sharing awareness data. The advantage is that it automatically detects when users are offline and that it does not leave any traces in the Yjs document (in the log of changes). You can also have a look at the other editor bindings to see how they make use of the Awareness state. E.g. [y-codemirror](https://github.com/yjs/y-codemirror/blob/master/src/y-codemirror.js)

Reply:
Ahh, got it. So I would use `setLocalStateField` to set the cursors and `getStates` to read them?

Reply:
Exactly!